### DISCUSSION TITLE
Newbie need help with getStaticPaths

### DISCUSSION THREAD

**Discussion by 51ned** [vercel/next.js#30234] [Help]
Hi there and sorry for my English. I need help with [my project](https://github.com/51ned/nemesis/tree/base-settings-pt3). 
All page's data fetches from Contentful CMS uses GraphQL API. I tested common for all pages components in `index.js` - everything works fine.
In `pages/ekspertizy/avtotekhnichiskie/[slug].js`, when i try get slugs, like it written in [official documentation](https://nextjs.org/learn/basics/dynamic-routes/implement-getstaticpaths), i have '404 error' (see `getSlugs` function in `lib/service-page`). But there is no '404 error', if i getting slugs, like it written in [official example with Contentful CMS](https://github.com/vercel/next.js/blob/canary/examples/cms-contentful/pages/posts/%5Bslug%5D.js). Now i use 2-nd example, but there is no data rendered in `Page` function!
Function `getData` in 'lib/service-page', called inside `getStaticProps` and `getStaticPaths` functions in 'pages/ekspertizy/avtotekhnichiskie/[slug].js'  returns an array of objects, that looks like this:
```
[
  {
    title: someTitle,
    slug: someSlug
  },
 {
    title: someTitle,
    ...
  }
]
```
So where i'm wrong?

**Comment by ifonefox** [comment]
Is `getData` supposed to return something different if it has an argument? Right now, [it looks like it always returns an array with every page's data](https://github.com/51ned/nemesis/blob/base-settings-pt3/lib/service-page.js#L4)

**Reply by 51ned** [reply]
Yes, it returns an array with every page's data, cause i want to use dynamic routes to render all this pages.

**Reply by ifonefox** [reply]
I mean in getStaticProps, not getStaticPaths. It calls `getData(params.slug)`, which looks like it returns the same thing as `getData()`.

**Reply by ifonefox** [reply]
Also, your `getSlugs()` function doesn't return the data in the right format. It shouldn't return a slash before `i.slug`

Edit: `transformData` is undefined. Maybe that's why it didn't work

**Reply by 51ned** [reply]
> I mean in getStaticProps, not getStaticPaths. It calls `getData(params.slug)`, which looks like it returns the same thing as `getData()`.
Yep, my question in this. What should I call in getStaticProps? How should getData() look like, or what argument should I pass to it?

**Reply by ifonefox** [reply]
You should call a function that takes a single slug, and returns just the props for that page. Something like your existing `getData`, but instead of making an array of `data[i].head`s, it just return one page's

**Reply by 51ned** [reply]
But i want to use dynamic routes, as it described [here](https://nextjs.org/learn/basics/dynamic-routes/implement-getstaticprops):
```
export async function getStaticProps({ params }) {
  const postData = getPostData(params.id)
  return {
    props: {
      postData
    }
  }
}
```
In this example, getPostData in getStaticProps returns data for each of the two posts. And getStaticPaths returns an arr of slugs for this posts. As a result, we have two static pages, the data of each of them corresponds to their slugs.

In my project something about 30 pages on start. I can hardcode them all, but i can do this and without NextJS. :)

**Reply by ifonefox** [reply]
> getStaticProps returns data for each of the two posts

Nextjs runs `getStaticPaths` to get the list of pages it needs to render, then runs `getStaticProps` once for each set of params that `getStaticPaths` returns. `getStaticProps` doesn't return the data for every page all at once. On the page you just linked, it just returns the data for the one post with an ID that matches `params.id`. In your case, it would just return the data for the page with the slug that matches `params.slug`.

**Reply by 51ned** [reply]
Do I understand correctly that for this I need to add for my GraphQL-query something like this (taken from [here](https://github.com/vercel/next.js/blob/canary/examples/cms-contentful/lib/api.js)):
```
`query {
  postCollection(where: { slug: "${slug}" }) {
    slug
    ...
  }
}`,
```
Or fragment `(where: { slug: "${slug}")` here for another reason?
Anyway, thx for help one more time. Something starts to clear up in my thoughts.
